### PR TITLE
MTDSA-2180 : Integrated validator and controller

### PR DESCRIPTION
**Pull request by narendravijayarao**

### CODE REVIEW

File: app/v2/models/errors/SubmitEopsDeclarationErrors.scala [hunk 1]
```diff
@@ -23,9 +23,10 @@ object SubmitEopsDeclarationErrors {
 
   object MissingEndDateError extends Error("MISSING_END_DATE", "Missing account period end date")
   object InvalidEndDateError extends Error("FORMAT_END_DATE", "Invalid account period end date")
-
   object InvalidRangeError extends Error("RANGE_INVALID", "The start date must be the same day or before the from date")
 
+  object NotFinalisedDeclaration extends Error("INVALID_REQUEST", "The statement cannot be accepted without a declaration it is finalised.")
```
[the remark this answers is not in the document]
Comment (narendravijayarao) [reply app/v2/models/errors/SubmitEopsDeclarationErrors.scala]: instead what about this "The statement cannot be accepted without a declaration is finalised."

File: test/v2/controllers/EopsDeclarationControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            LocalDate.parse(from), LocalDate.parse(to))))
+
+        MockedEopsDeclarationService.submitDeclaration(EopsDeclarationSubmission(Nino(nino),
+          LocalDate.parse(from), LocalDate.parse(to)))
+          .returns(Future.successful(None))
+
+        private val response: Future[Result] =
+          testController.submit(nino, from, to)(fakePostRequest[JsValue](Json.parse(requestJson)))
+
+        status(response) shouldBe NO_CONTENT
```
[the remark this answers is not in the document]
Comment (narendravijayarao) [reply test/v2/controllers/EopsDeclarationControllerSpec.scala]: In success scenarios you won't get any body... We are covering that in IT tests, we can add here also if required.

File: project/plugins.sbt [hunk 3]
```diff
@@ -1,19 +1,3 @@
-/*
```
Comment (macgrewal) [inline project/plugins.sbt]: Please restore header

File: test/v2/services/EopsDeclarationServiceSpec.scala [hunk 4]
```diff
@@ -108,24 +121,19 @@ class EopsDeclarationServiceSpec extends ServiceSpec {
...
-              .returns(None)
-            MockedDesConnector.submitEOPSDeclaration(nino, from, to)
+            MockedDesConnector.submitEOPSDeclaration(Nino(nino), LocalDate.parse(start), LocalDate.parse(end))
               .returns(error)
 
-            val result: Option[ErrorResponse]  = await(service.submit(nino, from, to))
+            val result: Option[ErrorResponse]  = await(service.submit(EopsDeclarationSubmission(Nino(nino),
+              LocalDate.parse(start), LocalDate.parse(end))))
+
             result.get shouldBe ErrorResponse(mtdError, None)
```
Comment (macgrewal) [inline test/v2/services/EopsDeclarationServiceSpec.scala]: Assertion should be `result shouldBe Some(ErrorResponse(mtdError, None))`. If the value of `result` were `None` `result.get` would throw an error

File: test/v2/services/EopsDeclarationServiceSpec.scala [hunk 5]
```diff
@@ -16,77 +16,90 @@
...
+        MockedDesConnector.submitEOPSDeclaration(Nino(nino), LocalDate.parse(start), LocalDate.parse(end))
           .returns(Future{Some(desResponse)})
 
         val expected = ErrorResponse(BVRError, Some(Seq(RuleClass4Over16,RuleClass4PensionAge)))
 
-        val result: Option[ErrorResponse] = await(service.submit(nino, start, end))
+        val result: Option[ErrorResponse] = await(service.submit(EopsDeclarationSubmission(Nino(nino),
+          LocalDate.parse(start), LocalDate.parse(end))))
+
         result.get shouldBe expected
```
Comment (macgrewal) [inline test/v2/services/EopsDeclarationServiceSpec.scala]: Assertion should be `result shouldBe Some(expected)`. If the value of `result` were `None` `result.get` would throw an error

File: test/v2/services/EopsDeclarationServiceSpec.scala [hunk 6]
```diff
@@ -16,77 +16,90 @@
...
+
+        MockedDesConnector.submitEOPSDeclaration(Nino(nino), LocalDate.parse(start), LocalDate.parse(end))
+          .returns(Future{Some(desResponse)})
+
+        val expected = ErrorResponse(DownstreamError, None)
+
+        val result: Option[ErrorResponse] = await(service.submit(EopsDeclarationSubmission(Nino(nino),
+          LocalDate.parse(start), LocalDate.parse(end))))
+
         result.get shouldBe expected
```
Comment (macgrewal) [inline test/v2/services/EopsDeclarationServiceSpec.scala]: Assertion should be `result shouldBe Some(expected)`. If the value of `result` were `None` `result.get` would throw an error

File: test/v2/services/EopsDeclarationServiceSpec.scala [hunk 7]
```diff
@@ -16,77 +16,90 @@
...
-        MockedDesConnector.submitEOPSDeclaration(nino, start, end)
+        MockedDesConnector.submitEOPSDeclaration(Nino(nino), LocalDate.parse(start), LocalDate.parse(end))
           .returns(Future{Some(desResponse)})
 
         val expected = ErrorResponse(BadRequestError, Some(Seq(InvalidEndDateError,InvalidStartDateError)))
 
-        val result: Option[ErrorResponse] = await(service.submit(nino, start, end))
+        val result: Option[ErrorResponse] = await(service.submit(EopsDeclarationSubmission(Nino(nino),
+          LocalDate.parse(start), LocalDate.parse(end))))
+        result.get shouldBe expected
```
Comment (macgrewal) [inline test/v2/services/EopsDeclarationServiceSpec.scala]: Assertion should be `result shouldBe Some(expected)`. If the value of `result` were `None` `result.get` would throw an error

File: test/v2/services/EopsDeclarationServiceSpec.scala [hunk 8]
```diff
@@ -108,24 +121,19 @@ class EopsDeclarationServiceSpec extends ServiceSpec {
...
 
-            val result: Option[ErrorResponse]  = await(service.submit(nino, from, to))
+            val result: Option[ErrorResponse]  = await(service.submit(EopsDeclarationSubmission(Nino(nino),
+              LocalDate.parse(start), LocalDate.parse(end))))
+
             result.get shouldBe ErrorResponse(mtdError, None)
           }
         }
     }
-
```
Comment (macgrewal) [inline test/v2/services/EopsDeclarationServiceSpec.scala]: ??

**Pull request merged**